### PR TITLE
Only send valid state codes down to Step Function

### PR DESCRIPTION
**Pull request by paulbrown1982**
## Why are you doing this?
We receive at around one AWS alarm per day where the value provided in the State field is incorrect, either in terms of what Zuora can understand (2-letter code or full name), and in the set that matches the provided country (e.g. providing Ontario when country is United States).

This change includes a better lookup function in `country.js` called `stateProvinceFromString` that will search for a state based on a provided string of 2-characters, the full state name, or a 3-letter code (which isn't a validly structured code in USA and CA but is in Australia). The function also ensures that it looks up in the correct set of states, respective to the provided country. If there is no match, it returns null.

I also updated the logic which determines whether a state code is required by Zuora. 
For all other instances, including Single, we will not pass on the state value unless it matches the country's set of states - this is a change from the existing logic that would pass-through whatever value was entered. We see so many unparsable or mismatched values in the DB or Zuora, that the state value not worth collecting if it does not map based on the `stateProvinceFromString` search function.

I also renamed the `updateState` function and `'UPDATE_STATE'` reducer matcher to make it clear it's about the country's state / province rather than anything React related. 
 

[**Trello Card**](https://trello.com/c/pK5SzRGy)

## Screenshots
No change to the UI other than an updated error message.

Before:
Screen Shot 2020-03-02 at 13 54 06

After:
Screen Shot 2020-03-02 at 13 53 57

### CODE REVIEW

File: support-frontend/assets/helpers/internationalisation/country.js [hunk 1]
```diff
@@ -421,20 +425,26 @@ export const isInStripePaymentRequestAllowedCountries = (country: IsoCountry) =>
   stripePaymentRequestAllowedCountries.includes(country);
 
 // ----- Functions ----- /
+function sateFromMap(l: string, states: {[p: string]: string}): ?StateProvince {
```
Comment (rupertbates) [inline support-frontend/assets/helpers/internationalisation/country.js]: should be `stateFromMap`?

Comment (paulbrown1982) [reply support-frontend/assets/helpers/internationalisation/country.js]: Whoops, yup will rename to `stateProvinceFromMap`.

File: support-frontend/assets/helpers/internationalisation/country.js [hunk 2]
```diff
@@ -421,20 +425,26 @@ export const isInStripePaymentRequestAllowedCountries = (country: IsoCountry) =>
...
+  const s = l.toUpperCase();
+  return states[s] ? s :
+    Object.keys(states).find(key => states[key].toUpperCase() === s || (s.length === 3 && s.startsWith(key)));
```
Comment (jlieb10) [inline support-frontend/assets/helpers/internationalisation/country.js]: Should we call `toUpperCase()` on `s` as well, so we know we are comparing like for like?

File: support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx [hunk 3]
```diff
@@ -263,23 +246,28 @@ function onPayment(
...
+      countryAndStateUpdateOk = true;
+    } else if (!props.stateOrProvince) {
+      logException('Missing address_state in payment request token and no state/province selected in the form');
     }
   }
 
-  const nameUpdateOk: boolean = props.stripeAccount !== 'ONE_OFF' ?
+  // If recurring, Salesforce needs a valid last name.
+  const nameUpdateOk: boolean = props.contributionType !== 'ONE_OFF' ?
     updatePayerName(paymentRequestData, props.updateFirstName, props.updateLastName) : true;
```
Comment (rupertbates) [inline support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]:
```suggestion
  const nameUpdateOk: boolean = props.contributionType === 'ONE_OFF' || updatePayerName(paymentRequestData, props.updateFirstName, props.updateLastName);
```

Comment (paulbrown1982) [reply support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: I did think of doing it that way, however it then inverts the logic to the comment above. If you feel it's better I'll do it this way round for both the `...Ok` variables?

Comment (rupertbates) [reply support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: I think it works better with the name of the variable and is simpler, maybe tweak the comment slightly if you think it's misleading?

File: support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx [hunk 4]
```diff
@@ -260,26 +243,29 @@ function onPayment(
...
+  let countryAndStateValueOk = props.contributionType === 'ONE_OFF' || !['US', 'CA'].includes(validatedCountryFromCard);
+  if (validatedCountryFromCard) {
+    props.updateBillingCountry(validatedCountryFromCard);
+    const validatedBillingStateOrProvinceFromCard: Option<StateProvince> =
+      stateProvinceFromString(validatedCountryFromCard, billingStateOrProvinceFromCard);
+    if (validatedBillingStateOrProvinceFromCard) {
+      props.updateStateOrProvince(validatedBillingStateOrProvinceFromCard);
+      countryAndStateValueOk = true;
+    } else if (!props.stateOrProvince) {
+      logException('Missing address_state in payment request token and no state/province selected in the form');
```
Comment (tomrf1) [inline support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: this error will be logged even if a state/province is not required, e.g. a UK user

Comment (paulbrown1982) [reply support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: Ah yes, before it only logged if US and Canada. I'll raise a fix.

Comment (paulbrown1982) [reply support-frontend/assets/pages/contributions-landing/components/StripePaymentRequestButton/StripePaymentRequestButton.jsx]: Fix is here: https://github.com/guardian/support-frontend/pull/2378

**Pull request merged**